core/src/main/scala/kafka/tools/GetOffsetShell.scala

```diff
@@ -140,22 +139,25 @@ object GetOffsetShell {
           Some(new TopicPartition(p.topic, p.partition))
       }
 
+      val offsetSpec = listOffsetsTimestamp match {
+        case ListOffsetsRequest.EARLIEST_TIMESTAMP => OffsetSpec.earliest()
+        case ListOffsetsRequest.LATEST_TIMESTAMP => OffsetSpec.latest()
+        case ListOffsetsRequest.MAX_TIMESTAMP => OffsetSpec.maxTimestamp()
+        case _ => OffsetSpec.forTimestamp(listOffsetsTimestamp)
+      }
+
+      val timestampsToSearch = topicPartitions.map(tp => tp -> offsetSpec).toMap.asJava
+
       /* Note that the value of the map can be null */
-      val partitionOffsets: collection.Map[TopicPartition, java.lang.Long] = listOffsetsTimestamp match {
-        case ListOffsetsRequest.EARLIEST_TIMESTAMP => consumer.beginningOffsets(topicPartitions.asJava).asScala
-        case ListOffsetsRequest.LATEST_TIMESTAMP => consumer.endOffsets(topicPartitions.asJava).asScala
-        case _ =>
-          val timestampsToSearch = topicPartitions.map(tp => tp -> (listOffsetsTimestamp: java.lang.Long)).toMap.asJava
-          consumer.offsetsForTimes(timestampsToSearch).asScala.map { case (k, x) =>
-            if (x == null) (k, null) else (k, x.offset: java.lang.Long)
-          }
+      val partitionOffsets = client.listOffsets(timestampsToSearch).all().get().asScala.map { case (k, x) =>
+        if (x == null) (k, null) else (k, x.offset: java.lang.Long)
```

Comment:
* Can `x` be `null` with the admin client?
* It might be better to iterate over the partitions and call `ListOffsetsResult.partitionResult` for each of them. If we do this, we can basically avoid filtering out the partitions without leaders as the admin client does it as well. Then, we can catch the relevant exception (`LEADER_NOT_AVAILABLE`) and log the error. That seems a bit better as the metadata used by the admin client are more up to date. This would also allow us to remove the conversion to `PartitionInfo` in `listPartitionInfos`. What do you think?

Reply:
Yes, this is a good idea, after moving the leader check to `ListOffsetsResult.partitionResult`, we can use `TopicPartition` in `TopicFilter`